rfd/0003-deduplication.md

```diff
@@ -0,0 +1,214 @@
+boefje should run. This is the serialized ooi, the result of calling `serialize`
+on an OOI.
+
+It might be better to provide the serialized ooi as `input_ooi` argument to the
+boefje `run` to make it more clear what the contents of the field is, see below
+for the proposed signature.
+
+#### Usage of organization field
+
+The external DB boefje currently needs the organization to fetch the data from
+the external database. Obviously this means the boefje can't be deduplicated.
+
+To support this we should add a flag to boefjes that require the organization.
+If the flag is set, we provide organization to the boefje and can't do
+deduplication. If the flag isn't set, we don't provide the organization and can
+duplicate. This way we make sure the organization is never used when it is not
+supposed to be.
+
+#### Usage of boefje task id field
+
+There is only one boefje that uses the task id. This
+
+The log4j vulnerability is now over 3 years old.
+
+For the sake of simplicity we might do away with passing the boefje the task id.
+
+#### Usage of input_ooi field
+
+Only the leakix boefje used the input_ooi field. For the sake of simplicy we
+should probably not provide this to the boefje. It should be easy to refactor
+this so it used the serialized input OOI.
+
+#### Usage of environment field
+
+This is not used directly from the BoefjeMeta, but this
+
+#### Usage of started_at, endated_at, boefje and runnable_hash fields
+
+This isn't used by any boefje and can be safely removed.
+
+#### Boefje signature / arguments
+
+Boefje that does not require organization can have the following signature:
+
+```python
+def run(input_ooi: dict[str, Any]) -> list[tuple[set, bytes | str]]:
+```
+
+And boefje that requires organization:
+
+```python
+def run(input_ooi: dict[str, Any], organization: str) -> list[tuple[set, bytes | str]]:
+```
+
+The `input_ooi` is the serialized input OOI that is stored in the BoefjeMeta
+`arguments`. The environment field does need to be passed as argument because it
+is already passed to the boefje using environment variables.
+
+### Deduplication implementation
+
+The scheduler is the place where boefjes are scheduled so that seems to be most
+obvious place to implement deduplication. Because the data necessary is
+currently spread out over multiple services / databases this means it would need
+to do requests to other services to fetch this information. This is unavoidable
+because there is currently no single place with all the information.
+
+#### Whether boefje needs organization
+
+The scheduler already fetches and caches all the available boefjes for an
+organization, so the scheduler can use this data to easily check if the boefje
+requires the organization and can't be deduplicated.
+
+#### Environment settings from previous task in other organization
+
+This data is stored in bytes and needs to be fetched by the scheduler to
+determine whether the environment settings are the same.
```

Comment:
Or, we could compare the runnable hash? Which should contain all these fields / settings.
The problem however is that The scheduler *can* look up the previous runnable hash, it cannot (since it cannot see the current boefje workers environment settings), compute the current runnable hash.

Reply:
Ah check. Then we should add a hash (we need that for our proofs anyway) I think?